web/website/content/posts/2023-03-14-pi-day.md

```diff
@@ -0,0 +1,370 @@
+only the rows where `n` is less than or equal to three. This produces `{"n":2}`
+in the next step and `{"n":3}` after that. On the next iteration `{"n":4}` is
+produced but that is eliminated by the filter condition. Since that leaves no
+new rows, iteration stops.
+
+If you try the query above in the
+[PRQL Playground](www.prql-lang.org/playground), the result set you get (in the
+"output.arrow" tab) is:
+
+| n   |
+| --- |
+| 1   |
+| 2   |
+| 3   |
+
+## Fibonacci Numbers
+
+Inspired by this, let's try to calculate Fibonnacci numbers which are often one
+of the first examples when recursion is introduced:
+
+```elm=
+from_text format:json '[{"a":1, "b":1}]'
+loop (
+    select [a=b, b=s"a"+b]
+)
+select a
+take 7
+```
+
+which produces the first 7 Fibonnaci numbers.
+
+| a   |
+| --- |
+| 1   |
+| 1   |
+| 2   |
+| 3   |
+
+|5 |8| |13|
+
+You might have noticed that we didn't actually include a `filter` in our loop
+this time. Instead we relied on the fact that DuckDB produces results lazily and
+since we only took 7 numbers it only produced what we needed.
+
+Now, let's set ourselves a bigger challenge.
+
+## Calculating the digits of pi
+
+March 14th is written on American calendars as 3/14 which reminds us of the
+first three digits of Pi=3.1415926535... . Therefore this day is commonly known
+as Pi day. In order to celebrate Pi Day 2023 and the recent release of `loop` in
+PRQL 0.6, why don't we try to calculate the digits of Pi in PRQL!
+
+For our query engine we will use [DuckDB](https://duckdb.org/) because it is
+really fast and has many modern features that make it ideally suited for the
+kinds of analytical queries that PRQL is targeting. There is even a
+[duckdb-prql](https://github.com/ywelsch/duckdb-prql) extension which allows you
+to write PRQL queries right inside DuckDB!
+
+We follow the algorithms in the paper
+["Unbounded Spigot Algorithms for the Digits of Pi" by Jeremy Gibbons](https://www.cs.ox.ac.uk/people/jeremy.gibbons/publications/spigot.pdf),
+in particular Rabinowitz and Wagon's Spigot Algorithm. While it is not the most
+efficient algorithm presented in the paper, it has the advantage that it uses
+only standard data types while the more efficient algorithms rely on unbounded
+integer types which are not available in many database systems.
+
+For my implementation, I adapted the Python implementation by John Burkardt
+found
+[here](https://people.sc.fsu.edu/~jburkardt/py_src/pi_spigot/pi_spigot.py). A
+bit of care had to be taken in porting the implementation to PRQL, as PRQL like
+SQL, is stateless so I introduced another state variable `k` to track the
+position inside the body of the loop.
+
+The resulting PRQL query is the following:
+
+```prql
+prql target:sql.duckdb
+
+let config = (
+    from_text format:json '[{"n":50}]'
+    derive [
+        len = (10*n)/3,
+        calc_steps = 1+4,
+        loop_steps = len + calc_steps,
+        ]
```

Comment:
```suggestion
    ]
```

Reply:
Not sure if it's worth the effort — hopefully soon we'll have an autoformatter... Lots of the brackets aren't aligned. I'm also generally using a 2-space indentation, but def don't do that manually!

Reply:
Yes, that would be helpful. I think I changed by alignment style halfway through so it's not consistent.